utils/frame/try-runtime/remote-externalities/src/lib.rs

```diff
@@ -0,0 +1,454 @@
+	pub at: Option<Hash>,
+	/// An optional cache file to WRITE to, not for reading. Not cached if set to `None`.
+	pub cache: Option<CacheConfig>,
+	/// The modules to scrape. If empty, entire chain state will be scraped.
+	pub modules: Vec<String>,
+}
+
+impl Default for OnlineConfig {
+	fn default() -> Self {
+		Self {
+			uri: "http://localhost:9933".into(),
+			at: None,
+			cache: None,
+			modules: Default::default(),
+		}
+	}
+}
+
+/// Configuration of the cache.
+#[derive(Clone)]
+pub struct CacheConfig {
+	// TODO: I could mix these two into one filed, but I think separate is better bc one can be
+	// configurable while one not.
+	/// File name.
+	pub name: String,
+	/// Base directory.
+	pub directory: String,
+}
+
+impl Default for CacheConfig {
+	fn default() -> Self {
+		Self { name: "CACHE".into(), directory: ".".into() }
+	}
+}
+
+impl CacheConfig {
+	fn path(&self) -> PathBuf {
+		Path::new(&self.directory).join(self.name.clone())
+	}
+}
+
+/// Builder for remote-externalities.
+pub struct Builder {
+	inject: Vec<KeyPair>,
+	mode: Mode,
+	chain: String,
+}
+
+impl Default for Builder {
+	fn default() -> Self {
+		Self {
+			inject: Default::default(),
+			mode: Mode::Online(OnlineConfig {
+				at: None,
+				uri: "http://localhost:9933".into(),
+				cache: None,
+				modules: Default::default(),
+			}),
+			chain: "UNSET".into(),
+		}
+	}
+}
+
+// Mode methods
+impl Builder {
+	fn as_online(&self) -> &OnlineConfig {
+		match &self.mode {
+			Mode::Online(config) => &config,
+			_ => panic!("Unexpected mode: Online"),
+		}
+	}
+
+	fn as_online_mut(&mut self) -> &mut OnlineConfig {
+		match &mut self.mode {
+			Mode::Online(config) => config,
+			_ => panic!("Unexpected mode: Online"),
+		}
+	}
+}
+
+// RPC methods
+impl Builder {
+	async fn rpc_get_head(&self) -> Hash {
+		let mut rt = tokio::runtime::Runtime::new().expect("Unable to create a runtime");
+		let uri = self.as_online().uri.clone();
+		rt.block_on::<_, _, ()>(futures::lazy(move || {
+			trace!(target: LOG_TARGET, "rpc: finalized_head");
+			let client: sc_rpc_api::chain::ChainClient<Number, Hash, (), ()> =
+				jsonrpc_core_client::transports::http::connect(&uri).wait().unwrap();
+			Ok(client.finalized_head().wait().unwrap())
+		}))
+		.unwrap()
+	}
+
+	/// Relay the request to `state_getPairs` rpc endpoint.
+	///
+	/// Note that this is an unsafe RPC.
+	async fn rpc_get_pairs(&self, prefix: StorageKey, at: Hash) -> Vec<KeyPair> {
+		let mut rt = tokio::runtime::Runtime::new().expect("Unable to create a runtime");
+		let uri = self.as_online().uri.clone();
+		rt.block_on::<_, _, ()>(futures::lazy(move || {
+			trace!(target: LOG_TARGET, "rpc: storage_pairs: {:?} / {:?}", prefix, at);
+			let client: sc_rpc_api::state::StateClient<Hash> =
+				jsonrpc_core_client::transports::http::connect(&uri).wait().unwrap();
+			Ok(client.storage_pairs(prefix, Some(at)).wait().unwrap())
+		}))
+		.unwrap()
+	}
```

Comment:
Why is there another async runtime here? You couldn't just use await so it goes to your first runtime??

Reply:
I believe we talked about this in chat :D The tokio versions mismatch, I need a separate runtime here.

Reply:
also, in either case, I want this to be replaced with jsonrpsee soon, see: https://github.com/paritytech/substrate/pull/8105

Reply:
Ohhh right I see! But then do you need the first runtime created by the runner? Oh wait I see, you don't want to change the API later on to async again when it's fixed.

Ok ok... so that's some "temporary" code but not much I guess. Approved.

Reply:
Could the `.compat()` help with that?

Reply:
I think I can't get rid of the upper one because I need some async closure to execute the `async fn run` in it. Or maybe I am wrong.

Reply:
https://github.com/tokio-rs/tokio-compat

Reply:
You can use `sync_run` instead of `async_run` if you don't need the runtime. But you're gonna need to change it back as soon as you finish #8105 so I wouldn't advice for it

Reply:
The Runner is a helper to create the `Configuration` and it can finish by: running async code (async_run), running sync code (sync_run) and run a substrate node (run_node_until_exit)

Reply:
(If things work and you're going to change it soon anyway: maybe dont bother too much on this PR)

Reply:
Given that I hope this will soon change, shall we let the nested runtime for now then?

Reply:
Yes that's what i meant